File: backend/app/schemas/maintenance_collection_reminders.py
r"""回款提醒 API v1 DTO（K0 Task 1 Step 1.5）。

逐项镜像 ``.ai/contracts/maintenance-collections/collection-reminders-api-v1.yaml``：
- 所有请求 schema 使用 ``ConfigDict(extra="forbid")``；
- 金额字段只接受十进制定点字符串或 null（``StrictStr`` 拒绝 Pydantic v2 对
  int/float 的宽松 str 转换），形状 ``^[0-9]+(?:\.[0-9]{1,2})?$`` 与 XLS 合同一致；
- follow-up 判别规则：handle 只允许 note；reschedule 必填 planned_month+reason
  且禁 note；reopen 必填 reason 且禁 planned_month+note；
- 错误体：domain_error（code/message/current_version/current_data_version/issues[]）
  与 request_validation_error（detail: loc/msg/type）。
source-file 端点是二进制附件响应，无 JSON DTO，不在此模块。
"""

from __future__ import annotations

from datetime import date, datetime
from typing import Annotated, Literal

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    StrictStr,
    StringConstraints,
    field_validator,
    model_validator,
)

# 金额文本：只接受十进制定点字符串（XLS 合同 plan_amount.text_regex）。
DecimalAmount = Annotated[
    StrictStr,
    StringConstraints(pattern=r"^[0-9]+(?:\.[0-9]{1,2})?$"),
]
# 计划月份：固定 YYYY-MM，月份 01–12。
YearMonth = Annotated[
    StrictStr,
    StringConstraints(pattern=r"^\d{4}-(0[1-9]|1[0-2])$"),
]

DatePrecision = Literal["day", "month"]
FollowUpStatus = Literal["pending", "handled"]
ReminderState = Literal[
    "needs_review", "handled", "incomplete", "overdue", "due_this_month", "upcoming"
]
FollowUpAction = Literal["handle", "reschedule", "reopen"]
OwnerScope = Literal["me", "all"]
AmountVisibility = Literal["visible", "restricted"]
BatchStatus = Literal["valid", "error", "applied", "expired"]
BindingStatus = Literal["reviewed"]
PreviewBindingStatus = Literal["reviewed", "pending_review"]
MilestoneChangeKind = Literal["create", "update", "unchanged", "source_missing"]
IssueSeverity = Literal["warning", "blocker"]


# ---------- 通用引用结构 ----------
class ManagerAssignment(BaseModel):
    username: str | None
    display_name: str | None


class ServicePeriod(BaseModel):
    service_start: date | None
    service_end: date | None
    completeness_state: str


class ContractRef(BaseModel):
    project_contract_id: str
    contract_no: str | None
    relation_status: str
    lifecycle_status: str
    version: int = Field(ge=1)


class ProjectRef(BaseModel):
    project_id: str
    project_code: str
    display_name: str
    lifecycle_status: str
    version: int = Field(ge=1)
    manager_assignment: ManagerAssignment
    service_period: ServicePeriod
    contracts: list[ContractRef]


class ReminderCounts(BaseModel):
    total: int = Field(ge=0)
    needs_review: int = Field(ge=0)
    handled: int = Field(ge=0)
    incomplete: int = Field(ge=0)
    overdue: int = Field(ge=0)
    due_this_month: int = Field(ge=0)
    upcoming: int = Field(ge=0)


class ActionableMilestone(BaseModel):
    milestone_id: str
    project_contract_id: str
    contract_no: str | None
    sequence: int = Field(ge=1, le=24)
    planned_month: YearMonth | None
    planned_amount: DecimalAmount | None
    reminder_state: ReminderState
    version: int = Field(ge=1)


class MilestoneOperation(BaseModel):
    operation_id: str
    action: FollowUpAction
    reason: str | None
    actor_display_name: str
    created_at: datetime
    result_version: int = Field(ge=1)


class MilestoneRow(BaseModel):
    milestone_id: str
    project_contract_id: str
    contract_no: str | None
    sequence: int = Field(ge=1, le=24)
    planned_date: date | None
    date_precision: DatePrecision
    planned_month: YearMonth | None
    planned_amount: DecimalAmount | None
    completeness_state: str
    follow_up_status: FollowUpStatus
    reminder_state: ReminderState
    follow_up_review_required: bool
    followed_up_by: str | None
    followed_up_at: datetime | None
    follow_up_note: str | None
    last_operation: MilestoneOperation | None
    version: int = Field(ge=1)


# ---------- 目录 / 详情 ----------
class DirectoryRow(BaseModel):
    project: ProjectRef
    reminder_counts: ReminderCounts
    next_actionable_milestone: ActionableMilestone | None


class DirectoryResponse(BaseModel):
    rows: list[DirectoryRow]
    total: int = Field(ge=0)
    page: int = Field(ge=1)
    page_size: int = Field(ge=1, le=200)
    owner_scope: OwnerScope
    allowed_owner_scopes: list[OwnerScope]
    as_of: date
    data_version: str
    amount_visibility: AmountVisibility

    @field_validator("allowed_owner_scopes")
    @classmethod
    def _owner_scopes_unique(cls, value: list[OwnerScope]) -> list[OwnerScope]:
        """allowed_owner_scopes 必须去重（YAML ``unique: true``，P1-2）。"""
        if len(value) != len(set(value)):
            raise ValueError("allowed_owner_scopes 不允许重复")
        return value

    @model_validator(mode="after")
    def _amount_visibility_fail_closed(self) -> "DirectoryResponse":
        """金额受限时目录里所有计划金额必须为 null（设计 §7.1，P1-2）：
        不能只靠前端隐藏。"""
        if self.amount_visibility == "restricted":
            for row in self.rows:
                milestone = row.next_actionable_milestone
                if milestone is not None and milestone.planned_amount is not None:
                    raise ValueError("金额受限时所有计划金额必须为空")
        return self


class ProjectDetailResponse(BaseModel):
    project: ProjectRef
    summary: ReminderCounts
    rows: list[MilestoneRow]
    as_of: date
    data_version: str
    amount_visibility: AmountVisibility

    @model_validator(mode="after")
    def _amount_visibility_fail_closed(self) -> "ProjectDetailResponse":
        """金额受限时详情里任何行的计划金额必须为 null（设计 §7.1，P1-2）。"""
        if self.amount_visibility == "restricted":
            for row in self.rows:
                if row.planned_amount is not None:
                    raise ValueError("金额受限时所有计划金额必须为空")
        return self


# ---------- follow-up ----------
class FollowUpRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    expected_version: int = Field(ge=1)
    idempotency_key: str = Field(min_length=8, max_length=128)
    action: FollowUpAction
    planned_month: YearMonth | None = None
    note: str | None = Field(default=None, max_length=1000)
    reason: str | None = Field(default=None, max_length=1000)

    @model_validator(mode="after")
    def _discriminated_fields(self) -> "FollowUpRequest":
        """判别规则：handle 只允许 note；reschedule 必填 planned_month+reason 且禁
        note；reopen 必填 reason 且禁 planned_month+note（YAML discriminated_rules）。"""
        if self.action == "handle":
            if self.planned_month is not None or self.reason is not None:
                raise ValueError("标记已处理只允许填写备注，不允许月份或理由")
        elif self.action == "reschedule":
            if self.planned_month is None or self.reason is None:
                raise ValueError("改期必须同时提供计划月份与理由")
            if not self.reason.strip():
                raise ValueError("改期理由去除首尾空白后必须非空")
            if self.note is not None:
                raise ValueError("改期不允许填写备注")
        elif self.action == "reopen":
            if self.reason is None or not self.reason.strip():
                raise ValueError("重新打开必须提供去除首尾空白后非空的理由")
            if self.planned_month is not None or self.note is not None:
                raise ValueError("重新打开不允许填写月份或备注")
        return self


class FollowUpResponse(BaseModel):
    row: MilestoneRow
    data_version: str
    idempotent_replay: bool


# ---------- 导入预览 ----------
class ImportIssue(BaseModel):
    code: str
    severity: IssueSeverity
    row_key: str | None
    sequence: int | None = Field(ge=1, le=24)
    message: str


class MilestoneDiff(BaseModel):
    sequence: int = Field(ge=1, le=24)
    planned_month: YearMonth | None
    planned_amount: DecimalAmount | None
    change: MilestoneChangeKind
    expected_milestone_version: int | None = Field(ge=1)


class PreviewBinding(BaseModel):
    status: PreviewBindingStatus
    project_id: str | None
    project_version: int | None = Field(ge=1)
    project_contract_id: str | None
    project_contract_version: int | None = Field(ge=1)
    existing_binding_version: int | None = Field(ge=1)


class PreviewOrder(BaseModel):
    row_key: str
    external_order_no: str
    source_project_name: str | None
    binding: PreviewBinding
    milestone_diffs: list[MilestoneDiff]
    warning_codes: list[str]
    blocker_codes: list[str]


class PreviewCounts(BaseModel):
    projects: int = Field(ge=0)
    milestones: int = Field(ge=0)
    bound: int = Field(ge=0)
    pending_binding: int = Field(ge=0)
    blockers: int = Field(ge=0)
    warnings: int = Field(ge=0)
    create: int = Field(ge=0)
    update: int = Field(ge=0)
    unchanged: int = Field(ge=0)
    source_missing: int = Field(ge=0)


class PreviewResponse(BaseModel):
    batch_id: str
    batch_version: int = Field(ge=1)
    data_version: str
    status: BatchStatus
    contract_version: str
    file_sha256: str = Field(pattern=r"^[0-9a-f]{64}$")
    counts: PreviewCounts
    rows: list[PreviewOrder]
    issues: list[ImportIssue]
    can_apply: bool
    expires_at: datetime


# ---------- 绑定候选搜索 ----------
class BindingOptionContract(BaseModel):
    project_contract_id: str
    contract_no: str | None
    relation_status: str
    lifecycle_status: str
    version: int = Field(ge=1)


class BindingOptionProject(BaseModel):
    project_id: str
    project_code: str
    display_name: str
    version: int = Field(ge=1)
    contracts: list[BindingOptionContract]


class BindingOptionsResponse(BaseModel):
    batch_id: str
    rows: list[BindingOptionProject]
    total: int = Field(ge=0)
    page: int = Field(ge=1)
    page_size: int = Field(ge=1, le=50)
    q: str


# ---------- 应用 ----------
class ApplyBinding(BaseModel):
    model_config = ConfigDict(extra="forbid")

    row_key: str
    external_order_no: str
    project_id: str
    project_version: int = Field(ge=1)
    project_contract_id: str
    project_contract_version: int = Field(ge=1)
    existing_binding_version: int | None = Field(ge=1)
    reason: str | None = Field(max_length=1000)

    @model_validator(mode="after")
    def _reassignment_requires_reason(self) -> "ApplyBinding":
        """改派已有绑定必须填写非空理由；新绑定 existing_binding_version 必须为 null。"""
        if self.existing_binding_version is not None and (
            self.reason is None or not self.reason.strip()
        ):
            raise ValueError("改派已有绑定必须填写理由")
        return self


class ApplyRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    expected_batch_version: int = Field(ge=1)
    expected_data_version: str
    bindings: list[ApplyBinding]


class ApplyCounts(BaseModel):
    created: int = Field(ge=0)
    updated: int = Field(ge=0)
    unchanged: int = Field(ge=0)
    source_missing: int = Field(ge=0)
    needs_review: int = Field(ge=0)


class ApplyResponse(BaseModel):
    batch_id: str
    batch_version: int = Field(ge=1)
    data_version: str
    status: BatchStatus
    counts: ApplyCounts
    idempotent_replay: bool
    applied_at: datetime


# ---------- 搜索请求 ----------
class SearchRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    q: str = Field(default="", max_length=256)
    owner_scope: OwnerScope = "me"
    reminder_state: ReminderState | None = None
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=24, ge=1, le=200)


# ---------- 错误体（error_contract） ----------
class DomainErrorDetail(BaseModel):
    code: str
    message: str
    current_version: int | None = None
    current_data_version: str | None = None
    issues: list[ImportIssue] = Field(default_factory=list)


class DomainError(BaseModel):
    detail: DomainErrorDetail


class ValidationErrorItem(BaseModel):
    loc: list[str | int]
    msg: str
    type: str


class RequestValidationError(BaseModel):
    detail: list[ValidationErrorItem]

File: frontend/src/api/maintenanceCollectionReminders.ts
import { api } from "../api";

/**
 * 回款计划提醒专用 API（车道 C1）。
 *
 * 契约：.ai/contracts/maintenance-collections/collection-reminders-api-v1.yaml
 * 本模块只服务 reminder-only 口径，不得混入维护旧财务回款快照的
 * maintenanceOperations.ts，也不得把 planned_amount 转成 JavaScript number。
 *
 * 服务端完整路径带 /api 前缀，前端 axios baseURL=/api，这里只写 /maintenance/...
 */

export type CollectionDatePrecision = "day" | "month";
export type CollectionFollowUpStatus = "pending" | "handled";
export type CollectionReminderState =
  | "needs_review"
  | "handled"
  | "incomplete"
  | "overdue"
  | "due_this_month"
  | "upcoming";
export type CollectionFollowUpAction = "handle" | "reschedule" | "reopen";
export type CollectionOwnerScope = "me" | "all";
export type CollectionAmountVisibility = "visible" | "restricted";
export type CollectionBatchStatus = "valid" | "error" | "applied" | "expired";
export type CollectionBindingStatus = "reviewed";
export type CollectionPreviewBindingStatus = "reviewed" | "pending_review";
export type CollectionMilestoneChange =
  | "create"
  | "update"
  | "unchanged"
  | "source_missing";
export type CollectionIssueSeverity = "warning" | "blocker";

export interface CollectionManagerAssignment {
  username: string | null;
  display_name: string | null;
}

export interface CollectionServicePeriod {
  service_start: string | null;
  service_end: string | null;
  completeness_state: string;
}

export interface CollectionContractRef {
  project_contract_id: string;
  contract_no: string | null;
  relation_status: string;
  lifecycle_status: string;
  version: number;
}

export interface CollectionProjectRef {
  project_id: string;
  project_code: string;
  display_name: string;
  lifecycle_status: string;
  version: number;
  manager_assignment: CollectionManagerAssignment;
  service_period: CollectionServicePeriod;
  contracts: CollectionContractRef[];
}

export interface CollectionReminderCounts {
  total: number;
  needs_review: number;
  handled: number;
  incomplete: number;
  overdue: number;
  due_this_month: number;
  upcoming: number;
}

export interface CollectionActionableMilestone {
  milestone_id: string;
  project_contract_id: string;
  contract_no: string | null;
  sequence: number;
  planned_month: string | null;
  planned_amount: string | null;
  reminder_state: CollectionReminderState;
  version: number;
}

export interface CollectionDirectoryRow {
  project: CollectionProjectRef;
  reminder_counts: CollectionReminderCounts;
  next_actionable_milestone: CollectionActionableMilestone | null;
}

export interface CollectionReminderDirectoryResponse {
  rows: CollectionDirectoryRow[];
  total: number;
  page: number;
  page_size: number;
  owner_scope: CollectionOwnerScope;
  allowed_owner_scopes: CollectionOwnerScope[];
  as_of: string;
  data_version: string;
  amount_visibility: CollectionAmountVisibility;
}

export interface CollectionMilestoneOperation {
  operation_id: string;
  action: CollectionFollowUpAction;
  reason: string | null;
  actor_display_name: string;
  created_at: string;
  result_version: number;
}

export interface CollectionMilestoneRow {
  milestone_id: string;
  project_contract_id: string;
  contract_no: string | null;
  sequence: number;
  planned_date: string | null;
  date_precision: CollectionDatePrecision;
  planned_month: string | null;
  planned_amount: string | null;
  completeness_state: string;
  follow_up_status: CollectionFollowUpStatus;
  reminder_state: CollectionReminderState;
  follow_up_review_required: boolean;
  followed_up_by: string | null;
  followed_up_at: string | null;
  follow_up_note: string | null;
  last_operation: CollectionMilestoneOperation | null;
  version: number;
}

export interface CollectionProjectDetailResponse {
  project: CollectionProjectRef;
  summary: CollectionReminderCounts;
  rows: CollectionMilestoneRow[];
  as_of: string;
  data_version: string;
  amount_visibility: CollectionAmountVisibility;
}

export interface CollectionFollowUpRequest {
  expected_version: number;
  idempotency_key: string;
  action: CollectionFollowUpAction;
  planned_month?: string | null;
  note?: string | null;
  reason?: string | null;
}

export interface CollectionFollowUpResponse {
  row: CollectionMilestoneRow;
  data_version: string;
  idempotent_replay: boolean;
}

export interface CollectionImportIssue {
  code: string;
  severity: CollectionIssueSeverity;
  row_key: string | null;
  sequence: number | null;
  message: string;
}

export interface CollectionMilestoneDiff {
  sequence: number;
  planned_month: string | null;
  planned_amount: string | null;
  change: CollectionMilestoneChange;
  expected_milestone_version: number | null;
}

export interface CollectionPreviewBinding {
  status: CollectionPreviewBindingStatus;
  project_id: string | null;
  project_version: number | null;
  project_contract_id: string | null;
  project_contract_version: number | null;
  existing_binding_version: number | null;
}

export interface CollectionPreviewOrder {
  row_key: string;
  external_order_no: string;
  source_project_name: string | null;
  binding: CollectionPreviewBinding;
  milestone_diffs: CollectionMilestoneDiff[];
  warning_codes: string[];
  blocker_codes: string[];
}

export interface CollectionPreviewCounts {
  projects: number;
  milestones: number;
  bound: number;
  pending_binding: number;
  blockers: number;
  warnings: number;
  create: number;
  update: number;
  unchanged: number;
  source_missing: number;
}

export interface CollectionPreviewResponse {
  batch_id: string;
  batch_version: number;
  data_version: string;
  status: CollectionBatchStatus;
  contract_version: string;
  file_sha256: string;
  counts: CollectionPreviewCounts;
  rows: CollectionPreviewOrder[];
  issues: CollectionImportIssue[];
  can_apply: boolean;
  expires_at: string;
}

export interface CollectionBindingOptionContract {
  project_contract_id: string;
  contract_no: string | null;
  relation_status: string;
  lifecycle_status: string;
  version: number;
}

export interface CollectionBindingOptionProject {
  project_id: string;
  project_code: string;
  display_name: string;
  version: number;
  contracts: CollectionBindingOptionContract[];
}

export interface CollectionBindingOptionsResponse {
  batch_id: string;
  rows: CollectionBindingOptionProject[];
  total: number;
  page: number;
  page_size: number;
  q: string;
}

export interface CollectionApplyBinding {
  row_key: string;
  external_order_no: string;
  project_id: string;
  project_version: number;
  project_contract_id: string;
  project_contract_version: number;
  existing_binding_version: number | null;
  reason: string | null;
}

export interface CollectionApplyRequest {
  expected_batch_version: number;
  expected_data_version: string;
  bindings: CollectionApplyBinding[];
}

export interface CollectionApplyCounts {
  created: number;
  updated: number;
  unchanged: number;
  source_missing: number;
  needs_review: number;
}

export interface CollectionApplyResponse {
  batch_id: string;
  batch_version: number;
  data_version: string;
  status: CollectionBatchStatus;
  counts: CollectionApplyCounts;
  idempotent_replay: boolean;
  applied_at: string;
}

export interface CollectionReminderSearchRequest {
  q?: string;
  owner_scope?: CollectionOwnerScope;
  reminder_state?: CollectionReminderState | null;
  page?: number;
  page_size?: number;
}

const collectionMilestonesPath = (projectId: string) =>
  `/maintenance/projects/stable/${encodeURIComponent(projectId)}/collection-milestones`;

const collectionMilestoneBase = (milestoneId: string) =>
  `/maintenance/collection-milestones/${encodeURIComponent(milestoneId)}`;

const collectionPlanImportBase = (batchId: string) =>
  `/maintenance/collection-plan-imports/${encodeURIComponent(batchId)}`;

export const searchCollectionReminders = (
  body: CollectionReminderSearchRequest = {},
  options: { signal?: AbortSignal } = {},
) => {
  const request = {
    q: body.q?.trim() || "",
    owner_scope: body.owner_scope ?? "me",
    reminder_state: body.reminder_state ?? null,
    page: body.page ?? 1,
    page_size: body.page_size ?? 24,
  };
  return options.signal
    ? api.post<CollectionReminderDirectoryResponse>(
      "/maintenance/collection-reminders/search",
      request,
      { signal: options.signal },
    )
    : api.post<CollectionReminderDirectoryResponse>(
      "/maintenance/collection-reminders/search",
      request,
    );
};

export const getCollectionMilestones = (
  projectId: string,
  options: { signal?: AbortSignal } = {},
) => options.signal
  ? api.get<CollectionProjectDetailResponse>(collectionMilestonesPath(projectId), {
    signal: options.signal,
  })
  : api.get<CollectionProjectDetailResponse>(collectionMilestonesPath(projectId));

export const followUpCollectionMilestone = (
  milestoneId: string,
  body: CollectionFollowUpRequest,
) => api.post<CollectionFollowUpResponse>(
  `${collectionMilestoneBase(milestoneId)}/follow-ups`,
  body,
);

export const previewCollectionPlan = (file: File, idempotencyKey: string) => {
  const form = new FormData();
  form.append("file", file);
  return api.post<CollectionPreviewResponse>(
    "/maintenance/collection-plan-imports/preview",
    form,
    {
      headers: { "Idempotency-Key": idempotencyKey },
      timeout: 120000,
    },
  );
};

export const searchCollectionPlanBindingOptions = (
  batchId: string,
  params: { q: string; page?: number; page_size?: number },
  options: { signal?: AbortSignal } = {},
) => {
  const query = {
    q: params.q.trim(),
    page: params.page ?? 1,
    page_size: params.page_size ?? 20,
  };
  return options.signal
    ? api.get<CollectionBindingOptionsResponse>(
      `${collectionPlanImportBase(batchId)}/binding-options`,
      { params: query, signal: options.signal },
    )
    : api.get<CollectionBindingOptionsResponse>(
      `${collectionPlanImportBase(batchId)}/binding-options`,
      { params: query },
    );
};

export const applyCollectionPlan = (
  batchId: string,
  body: CollectionApplyRequest,
) => api.post<CollectionApplyResponse>(
  `${collectionPlanImportBase(batchId)}/apply`,
  body,
);

export const downloadCollectionPlanSourceFile = (batchId: string) =>
  api.get<Blob>(`${collectionPlanImportBase(batchId)}/source-file`, {
    responseType: "blob",
  });

/**
 * 金额展示格式化器：只接受十进制定点字符串或 null。
 *
 * - 只做合法十进制校验与千分位展示，不做任何 JavaScript 浮点计算；
 * - 不调用接收 number 的 money() 等既有金额 helper；
 * - null / 非法字符串返回 null，由调用方展示“无权限查看”等脱敏文案。
 */
export function formatDecimalAmount(value: string | null): string | null {
  if (value == null) return null;
  const normalized = value.trim();
  if (!/^[+-]?\d+(\.\d+)?$/.test(normalized)) return null;
  const negative = normalized.startsWith("-");
  const unsigned = normalized.replace(/^[+-]/, "");
  const [integerPart, fraction] = unsigned.split(".");
  const grouped = integerPart.replace(/\B(?=(\d{3})+(?!\d))/g, ",");
  return `${negative ? "-" : ""}${grouped}${fraction ? `.${fraction}` : ""}`;
}
